docs/getting-started/quickstart-programming.md

```diff
@@ -79,178 +62,227 @@ exports.up = function(knex) {
+You can add another GET API function to read all the greetings from the database like so:
+
+```javascript
+//export class Greetings {
+//...
+  @Transaction({readOnly: true})
+  @GetApi('/greetings')
+  static async allGreetings(ctxt: TransactionContext<Knex>) {
+    return await ctxt.client('greetings').select('*') as GreetingRecord[];
+  }
+//}
+```
+
+Here we use `@GetApi` and `@Transaction` together. This transaction only reads data so we mark it as `{readOnly: true}`. This enables DBOS to execute it faster, with fewer database round-trips.
 
 :::info
-In this quickstart, we write our database operations in raw SQL (using [knex.raw](https://knexjs.org/guide/raw.html)), but DBOS Transact also supports [knex's query builder](https://knexjs.org/guide/query-builder.html), [TypeORM](https://typeorm.io/), and [Prisma](https://www.prisma.io/docs/orm/prisma-client).
+In this quickstart, we run queries using [Knex query builder](https://knexjs.org/guide/query-builder.html). DBOS Transact also supports [Knex.raw](https://knexjs.org/guide/raw.html), [TypeORM](https://typeorm.io/), and [Prisma](https://www.prisma.io/docs/orm/prisma-client).
 :::
 
-### Interacting with External Services
+## Interacting with External Services
 
-:::info what you will learn
-How to safely send requests to third-party APIs.
-:::
+Now suppose we also want to record our greetings in a remote system. This could be something like sending an email or writing to some storage API. For this example, we'll use a demo DBOS Guestbook app. It lets us generate an API key and use it to record temporary greetings.
 
-Let's say we want to use a third-party client to send our greeting via e-mail.
-In DBOS, we strongly recommend wrapping calls to third-party APIs in [Communicators](../tutorials/communicator-tutorial).
-We'll see in the next section how communicators make your code more reliable.
-For now, add this code to `src/operations.ts`:
+### Create a Guestbook Key
+To create a new key, visit https://demo-guestbook.cloud.dbos.dev/key It should output a 36-character sequence like so `12345abc-1234-5678-1234-567890abcdef`. Yours will be different. You can pass this key to your app as a config variable. In your app folder, edit the file `dbos-config.yaml`. Add a new `env:` section at the bottom with the variable `GUESTBOOK_KEY` set to your key in quotes:
+```yaml
+env:
+  GUESTBOOK_KEY: 'your-key-value-here'
+```
+
+For example, if your key value is `12345abc-1234-5678-1234-567890abcdef` then you would add:
+```yaml
+env:
+  GUESTBOOK_KEY: '12345abc-1234-5678-1234-567890abcdef'
+```
+::::tip
+In production, to avoid storing keys in cleartext and to make rotation easier you could change this to `GUESTBOOK_KEY: ${ENV_GUESTBOOK_KEY}`. You would then set `ENV_GUESTBOOK_KEY` in your enviornment prior to starting or deploying the app. See the [Configuration Guide](../api-reference/configuration) for more details.
```

Comment:
```suggestion
In production, we recommend storing the key in an environment variable instead of plaintext. To do this, change the configuration to `GUESTBOOK_KEY: ${ENV_GUESTBOOK_KEY}` and set `ENV_GUESTBOOK_KEY` in your environment prior to starting or deploying the app. See the [Configuration Guide](../api-reference/configuration#environment-variables) for more details.
```